Applications/Spire/Source/Charting/ChartView.cpp

```diff
@@ -227,24 +282,36 @@ void ChartView::set_region(const ChartPoint& top_left,
   }
   m_top_left = top_left;
   m_bottom_right = bottom_right;
-  m_x_range = m_bottom_right.m_x - m_top_left.m_x;
-  m_x_axis_step = calculate_step(m_model->get_x_axis_type(), m_x_range);
-  m_y_range = m_top_left.m_y - m_bottom_right.m_y;
-  m_y_axis_step = calculate_step(m_model->get_y_axis_type(), m_y_range);
-  update_origins();
-  m_candlestick_promise = m_model->load(m_top_left.m_x, m_bottom_right.m_x);
-  m_candlestick_promise.then([=] (auto result) {
-    m_candlesticks = std::move(result.Get());
-    update_gaps();
+  auto bottom_right_pixel_x = m_bottom_right_pixel.x();
+  auto required_data = LoadedData{};
+  required_data.m_candlesticks = std::vector<Candlestick>();
+  required_data.m_gaps = std::vector<Gap>();
+  required_data.m_start = m_top_left.m_x;
+  required_data.m_end = m_bottom_right.m_x;
+  required_data.m_current_x = 0;
+  required_data.m_end_x = m_bottom_right_pixel.x();
+  required_data.m_values_per_pixel = (m_bottom_right.m_x - m_top_left.m_x) /
+    bottom_right_pixel_x;
+  m_loaded_data_promise = load_data(m_model->load(required_data.m_start,
+    required_data.m_end), required_data, m_model);
+  m_loaded_data_promise.then([=] (auto result) {
```

Comment:
This is a crash waiting to happen. ```then``` returns a new promise but you're not storing that promise anywhere.

Use:
```
m_loaded_data_promise = load_data(...).then(...);
```

Reply:
So should then(...) return a value or should I make it like:

m_loaded_data_promise = load_data(...);
m_loaded_data_promise = m_loaded_data_promise.then(...);

Reply:
```
m_loaded_data_promise = load_data(m_model->load(required_data.m_start,
    required_data.m_end), required_data, m_model).then([=] (auto result) {
  ...
});
```

Reply:
I get the compiler complaining that there's a mismatch of types:

 binary '=': no operator found which takes a right-hand operand of type 'void'

It appears to be evaluating the return value of then(...) as type: std::enable_if_t<true, void>

Reply:
Okay keep it as is for now, we'll revisit it later.